text/0580-destroyables.md

```diff
@@ -0,0 +1,163 @@
+of libraries, such as `ember-lifeline` and `ember-concurrency`, that would
+benefit from having a way to extend this hierarchy, adding their own "children"
+that are cleaned up whenever their parents are removed.
+
+Historically, Ember has exposed this cleanup lifecycle via _hooks_, such as the
+`willDestroy` hook on components. However, methods like these have a number of
+downsides:
+
+1. Since they are named, they can have collisions with other properties. This is
+   historically what led to the actions hash on classic components, in order to
+   avoid collisions between actions named `destroy` and the `destroy` lifecyle
+   hook.
+2. On a related note, relying on property names means that all framework classes
+   _must_ implement the `willDestroy` function (or another name), making it very
+   difficult to change APIs in the future.
+3. Methods are difficult for _libraries_ to instrument. For instance,
+   `ember-concurrency` currently replaces the `willDestroy` method on any class
+   with a task, with logic that looks similar to:
+
+   ```js
+   let PATCHED = new WeakSet();
+
+   function patchWillDestroy(obj) {
+     if (PATCHED.has(obj)) return;
+
+     let oldWillDestroy = obj.willDestroy;
+
+     obj.willDestroy = function() {
+       if (oldWillDestroy) oldWillDestroy.call(this);
+
+       teardownTasks(this);
+     }
+
+     PATCHED.add(obj);
+   }
+   ```
+
+   This logic becomes especially convoluted if _multiple_ libraries are
+   attempting to patch `willDestroy` in this way.
+
+4. Finally, since this isn't a standard, it's difficult to add _layers_ of new
+   destroyable values that can interoperate with one another. For instance,
+   there is no way for `ember-concurrency` to know how to destroy tasks on
+   non-framework classes that users may have added themselves.
+
+This RFC proposes a streamlined API that disconnects the exact implementation
+from any interface, allows for multiple destructors per-destroyable, and
+maximizes interoperability in general.
+
+## Detailed design
+
+The API consists of 3 functions:
+
+```ts
+declare function registerLifetime(parent: object, child: object): void;
+declare function registerDestructor(destroyable: object, destructor: () => void): void;
+declare function destroy(destroyable: object): void;
+```
+
+### `registerLifetime`
+
+This function is used to associate a destroyable object's lifetime with a
+parent. When the parent is destroyed, all registered children will also be
+destroyed.
+
+### `registerDestructor`
+
+Receives a destroyable and a destructor function, and associates the function
+with it. When the destroyable is destroyed, or when its parent is destroyed, the
+destructor function will be called. Multiple destructors can be associated with
+a given destroyable, and they can be associated over time, allowing libraries
+like `ember-lifeline` to dynamically add destructors as needed.
+
+### `destroy`
```

Comment:
Can you add prose to explain that calling `destroy(obj); destroy(obj);` is "safe" (e.g. calling destroy a second time is essentially inert